### ISSUE TITLE
Imports in `__init__.py`

### ISSUE THREAD

**Issue by csala** [HDI-Project/BTB#27]
**Note:** This issue arises from a presonal preference, so it is absolutely open to discussion

Even though importing objects from submodules in `__init__.py` to make them available at the root level is "pythonically accepted", I feel that this goes a bit against the "explicit better than implicit" statement in the **Zen of Python**.

Therefore, I suggest removing all the imports from the `__init__.py` modules and make all the import statements explicitly point at the source module.

**Comment by micahjsmith** [comment]
I mostly agree with your philosophy, I personally like to at least allow the most core functionality to be imported at the package level. That is, I think `from btb import GP` should be allowed, but `from btb import CLASS_GENERATOR` should definitely not be allowed. So perhaps these imports in `__init__.py` could be sharply reduced

**Comment by csala** [comment]
I'm afraid I disagree with this idea, especially given the nature of this project and having a mixture of functionalities among the submodules.

For example, how will then someone know whether `from btb import Uniform` is importing a tuner or a selector?

And this stays the same when limiting it to inside a submodule, like `btb.tuning` or `btb.selection`: since there are multiple classes per module, we can assume that each module is something like a "category". Then, when doing `from btb.tuning import XXX`, how do we know which category this `XXX` corresponds to?

Also, as the project grows, more and more classes will need to be put in the `__init__.py` file. Will we really need to import and load all the classes into memory if we only want to use one of them?

So, I see no difference between writing `from btb.tuning import GP` and `from btb.tuning.gp import GP`, but I feel that the second option provides several advantages, like explicitly informing about where to find the corresponding code and the nature of what is being imported, and allowing to precisely select what is  loaded into memory and what not.

**Comment by micahjsmith** [comment]
I think you make some good points.

Your point about ambiguous names ties into a suggestion I've previously made to rename `tuning.uniform.Uniform` to `tuning.uniform.UniformTuner` and `selection.uniform.Uniform` to `selection.uniform.UniformSelector` and so on. I'd argue that these unambiguous names for different components would allow someone to more easily import from the top level. Regardless, these names would clear up ambiguity about `from btb.tuning.uniform import Uniform` which is a paradigm that would otherwise cause confusion even if we did not reexport names in __init__ files.

Finally, I think that the main difference between `from btb import GP` and `from btb.tuning.gp import GP` is that an end-user does not have to have in-depth knowledge of package structure of btb for importing the most high-level, most commonly-used objects. For example, a pandas user finds it convenient to use `pd.Dataframe` instead of `pd.core.frame.DataFrame`, and probably 99% of pandas users would not know the full import path of DataFrame. This is not to say btb is nearly as complex as pandas, but I have never seen a package that requires an end-user to import 4 levels deep.

In summary, I think if we rename components to be less ambiguous names, and then re-export only *the most commonly used* components at the top level (`GPTuner`, `HyperParameter`, `ParamTypes`, maybe 2-3 other things), then we can provide a good user experience and not go over-board with maintaining different names, as you point out is a concern.

**Comment by csala** [comment]
The `pd.DataFrame` vs `pd.core.frame.DataFrame` was definitely a good example, and I can see how forcing the user to go 4 levels inside the module is not optimal.

At the same time, I would like to point out that this is actually a "core" component, very central to the rest of the project, and that in our case I think that the only objects that could play such a role are `HyperParameter` and `ParamTypes`.
For the other things, I think it would be better to keep them within their module branch: `btb.tuning`, `btb.selection` and so.

I basically understand it as in scikit-learn, where you import `sklearn.tree.DecisionTreeClassifier` or `sklearn.model_selection.train_test_split` instead of `sklearn.tree.tree.DecisionTreeClassifier` or `sklearn.model_selection._split.train_test_split`, but you cannot import `sklearn.DecisionTreeClassifier`.

On the other hand, if we make part of the objects available in the higher levels of the structure, I think that they should not be only _the most used ones_, but rather _the ones that we consider public_.
So, I would keep things like `CLASS_GENERATOR` or `BaseTuner` out of it, but include all the rest there.

So, to summarize, I would suggest making only these available:

```
from btb import HyperParameter, ParamTypes
from btb.tuning import (
    GCP, GCPEi, GCPEiVelocity, GP, GPEi, GPEiVelocity)
from btb.selection import (
    UCB1, Uniform, BestKReward...)  # All public selectors here.
```

**Comment by micahjsmith** [comment]
Those are more good points, and there is always a tradeoff.

I am on board with your suggestion! I continue to advocate for renaming `Uniform` -> `UniformTuner` etc. so that these objects can be used unambiguously in end-user applications.
